### PR TITLE
dnn(test): add YOLOv4 tests

### PR DESCRIPTION
**Pull request by alalek**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/762

relates #17148

<cut/>

```
force_builders=Linux32,Win32,Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.2.0:16.04
build_image:Custom Win=openvino-2020.2.0
build_image:Custom Mac=openvino-2020.2.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/perf/perf_net.cpp [hunk 1]
```diff
@@ -202,6 +202,18 @@ PERF_TEST_P_(DNNTestNetwork, YOLOv3)
     processNet("dnn/yolov3.weights", "dnn/yolov3.cfg", "", inp / 255);
 }
 
+PERF_TEST_P_(DNNTestNetwork, YOLOv4)
+{
+    if (backend == DNN_BACKEND_HALIDE)
+        throw SkipTestException("");
+    if (target == DNN_TARGET_MYRIAD)
+        throw SkipTestException("");
+    Mat sample = imread(findDataFile("dnn/dog416.png"));
```
Comment (dkurt) [inline modules/dnn/perf/perf_net.cpp:211]: Just to match origin behavior, we can also apply
```cpp
cvtColor(sample, sample, COLOR_BGR2RGB)
```
becase YOLOs work with RGB image.

Comment (alalek) [reply modules/dnn/perf/perf_net.cpp:212]: Fixed YOLOv3 too

File: modules/dnn/test/test_darknet_importer.cpp [hunk 2]
```diff
@@ -453,15 +463,24 @@ TEST_P(Test_Darknet_nets, YOLOv3)
...
-                                    1, 9,  0.785352f, 0.665503f, 0.373543f, 0.688893f, 0.439245f,  // a traffic light
-                                    1, 0,  0.980052f, 0.195856f, 0.378454f, 0.258626f, 0.629258f,  // a person
-                                    1, 2,  0.989633f, 0.450719f, 0.463353f, 0.496305f, 0.522258f,  // a car
-                                    1, 2,  0.997412f, 0.647584f, 0.459939f, 0.821038f, 0.663947f); // a car
+    const int N0 = 3;
+    const int N1 = 6;
+    static const float ref_[/* (N0 + N1) * 7 */] = {
+//0, 16, 0.998836f, 0.160024f, 0.389964f, 0.417885f, 0.943716f,
+//0, 1, 0.987908f, 0.150913f, 0.221933f, 0.742255f, 0.746261f,
+//0, 7, 0.952983f, 0.614621f, 0.150257f, 0.901368f, 0.289251f,
```
Comment (dkurt) [inline modules/dnn/test/test_darknet_importer.cpp:471]: Why do we need these commented lines?

Comment (alalek) [reply modules/dnn/test/test_darknet_importer.cpp:471]: They are dumped after NMS through code under `#if 0`
But test itself fails with them for some reason.

Comment (dkurt) [reply modules/dnn/test/test_darknet_importer.cpp:471]: But there are 3 lines with the same values below

Comment (alalek) [reply modules/dnn/test/test_darknet_importer.cpp]: You are right! They are duplicates from batch size 1/2.
Fixed.

**Pull request merged**